src/nds/nds_renderer.c

```diff
@@ -417,7 +417,7 @@ static void g_vtx(Gwords *words) {
     const Vtx *vertices = (const Vtx*)words->w1;
 
     // Store vertices in the vertex buffer
-    memcpy(&vertex_buffer[index - count], vertices, count * sizeof(Vtx));
+    swiFastCopy(vertices, &vertex_buffer[index - count], sizeof(Vtx) * 4);
```

Comment:
Why does this replace `count` with 4? That seems like a bug.

Reply:
fyi swiFastCopy is bugged as actually significantly slower than memcpy, if you want something faster try DMA or tonccpy or some other DS optimized memcpy

http://problemkaputt.de/gbatek-bios-memory-copy.htm

> BUG: The NDS/DSi uses the fast 32-byte-block processing only for the first N bytes (not for the first N words), so only the first quarter of the memory block is FAST, the remaining three quarters are SLOWLY copied word-by-word.

Reply:
> fyi swiFastCopy is bugged and actually significantly slower than memcpy, if you want something faster try DMA or tonccpy or some other DS optimized memcpy
> 
> http://problemkaputt.de/gbatek-bios-memory-copy.htm
> 
> > BUG: The NDS/DSi uses the fast 32-byte-block processing only for the first N bytes (not for the first N words), so only the first quarter of the memory block is FAST, the remaining three quarters are SLOWLY copied word-by-word.

Thank you for the info. I tried DMA copy and found it caused graphical corruption. This is the first I have heard of tonccpy. Would you say that normal swiCopy would be good for this?

Reply:
swiCopy is even worse than swiFastCopy, I did some testing and tonccpy seems to be more or less equal to memcpy (but VRAM safe), swiFastCopy is about 10% slower than memcpy, and swiCopy is about half the speed of memcpy. For whatever reason dmaCopy isn't cooperating with my testing so not sure exactly on it but I know it should be faster than memcpy.

You might need to [flush the cache](https://libnds.devkitpro.org/cache_8h.html#a4436939fe9b937948cebac4f7b74163a) for DMA copy to work, since DMA is separate from the CPU it can't access the CPU's cache. I'm not sure how big if a speed penalty cache flushing has so CPU caching might end up faster in some cases because of that.

Reply:
I would love to use dmaCopy but I can't seem to get it to work without corrupted graphics. I am flushing the cache.
I used
DC_FlushRange(&vertex_buffer[index - count], count * sizeof(Vtx));
dmaCopy(vertices, &vertex_buffer[index - count], count * sizeof(Vtx));
I have no idea what is wrong.

Reply:
You need to flush the source, not the destination

```c
DC_FlushRange(vertices, count * sizeof(Vtx));
```